### PR TITLE
chase: drop CHASE_EXTRACT_FILENAME and make chase_full() optionally return filename

### PR DESCRIPTION
**Pull request by yuwata**

### CODE REVIEW

File: src/libsystemd/sd-id128/id128-util.h
Comment (daandemeyer) [inline src/libsystemd/sd-id128/id128-util.h]: Why have two functions? Can't we just have one that calls `sd_id128_get_machine()` if `root == NULL`?

File: src/nspawn/nspawn.c
Comment (daandemeyer) [inline src/nspawn/nspawn.c]: This will now return -ENOMEDIUM when the machine ID is null, which triggers `ERRNO_IS_MACHINE_ID_UNSET()` which changes the behavior of this function

File: src/shared/specifier.c
Comment (daandemeyer) [inline src/shared/specifier.c]: I guess this translation was a copy paste bug?

File: src/shared/machine-id-setup.c
Comment (daandemeyer) [inline src/shared/machine-id-setup.c]: This will follow symlinks, which we shouldn't do here. I'd do a regular `chase_and_open()` with CHASE_NOFOLLOW here and then use `id128_read_fd()`

File: src/libsystemd/sd-id128/id128-util.c
Comment (daandemeyer) [inline src/libsystemd/sd-id128/id128-util.c]: Why open the parent here? We can open the file itself and use `id128_read_fd()` here.

I'd use a regular `chase()` here to get the file descriptor since `id128_read_at()` will reopen the fd for us anyway.

File: src/libsystemd/sd-id128/id128-util.c
Comment (daandemeyer) [inline src/libsystemd/sd-id128/id128-util.c]: The commit message for the commit that introduces this is wrong (says it introduces `id128_write_at()`

File: src/libsystemd/sd-id128/id128-util.c
Comment (daandemeyer) [inline src/libsystemd/sd-id128/id128-util.c]: I wonder if we should add `O_NOFOLLOW` here and require callers to chase() if they want to follow symlinks.

File: src/libsystemd/sd-id128/id128-util.c
Comment (daandemeyer) [inline src/libsystemd/sd-id128/id128-util.c]: Ditto, I wonder if it would make sense to use `O_NOFOLLOW` here as well.

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
Looks great! Some comments

File: src/libsystemd/sd-id128/id128-util.h
Comment (yuwata) [reply src/libsystemd/sd-id128/id128-util.h]: `sd_id128_get_machine()` caches the result. When `root` is specified, the result should not be cached.

File: src/nspawn/nspawn.c
Comment (yuwata) [reply src/nspawn/nspawn.c]: I think the behavior change should be desired. The change is now split out, and explain the behavior change in the commit message.

File: src/shared/specifier.c
Comment (yuwata) [reply src/shared/specifier.c]: Fixed.

File: src/libsystemd/sd-id128/id128-util.h
Comment (daandemeyer) [reply src/libsystemd/sd-id128/id128-util.h]: I was talking about id128_get_machine() and id128_get_machine_impl(). But I just saw that sd_id128_get_machine() now also calls id128_get_machine(), so it's done to avoid recursion.

Can you rename id128_get_machine_impl() to id128_get_machine_internal() for consistency with the rest of the codebase?

File: src/libsystemd/sd-id128/id128-util.h
Comment (daandemeyer) [inline src/libsystemd/sd-id128/id128-util.h]: Maybe we should make the flags enum name more generic since we don't have only formatting flags?

File: src/libsystemd/sd-id128/id128-util.h
Comment (daandemeyer) [inline src/libsystemd/sd-id128/id128-util.h]: When we open a file ...

File: src/libsystemd/sd-id128/id128-util.c
Comment (daandemeyer) [inline src/libsystemd/sd-id128/id128-util.c]: So regardless of whether the user sets ID128_NOFOLLOW when calling id128_write(), we've already resolved all symlinks we want to resolve at this point, so we should always add ID128_NOFOLLOW when calling id128_write_at() here so we definitely don't resolve any symlinks anymore.

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
Almost there

**Comment by yuwata** [comment]
@DaanDeMeyer Thank you for your review. I hope now all comments are addressed. I also added two small commits, one fixes an error handling in `chaseat()`, and the other introduces `id128_get_machine_at()`. PTAL.

File: src/libsystemd/sd-id128/id128-util.c
Comment (daandemeyer) [inline src/libsystemd/sd-id128/id128-util.c]: Now we have this weird inconsistency where `id128_read()` works on a root directory and chases symlinks and `id128_read_at()` works on a directory file descriptor and does not chase symlinks.

I would make this consistent by only having `id128_read()` take a path and not a root directory. and then have `id128_get_machine_id()` and `id128_get_machine_id_at()` which operate on a root directory so that the responsibilities are nicely separated. 

So `id128_read()` would be a classic `AT_FDCWD` wrapper of `id128_read_at()` like all the other ones we have.

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
Sorry for the many rounds, but I just realized an inconsistency that we should fix before merging

**Comment by yuwata** [comment]
@DaanDeMeyer Thank you for reviewing this so many times. Now, this PR becomes much bigger than I initially expected. But, I hope this is in a good shape now. PTAL!

File: src/basic/fd-util.h
Comment (daandemeyer) [inline src/basic/fd-util.h]: Commit message says "install dir_fd_is_root_or_cwd()", should be "introduce dir_fd_is_root_or_cwd()" I guess?

File: src/basic/fd-util.h
Comment (daandemeyer) [inline src/basic/fd-util.h]: Use `true` instead of `1` here?

File: src/basic/chase.c
Comment (daandemeyer) [inline src/basic/chase.c:576]: I'm curious when this matters, if `CHASE_PARENT` is set we'll always just reopen the file descriptor returned by `chase()`, and `chase()` will already ensure that the returned file descriptor points to a directory, so neither the open flags nor the mode will have an effect at that point? Or maybe I'm missing something here.

File: src/basic/chase.c
Comment (daandemeyer) [inline src/basic/chase.c:75]: Hmm, I understand the issue, but I wonder if it wouldn't be better solved by adding an optional `root` argument to `path_extract_filename()` which prevents extracting the filename of a directory outside of `root`.

That way we would avoid adding all the `full` variants of the `chase()` functions.

File: src/libsystemd/sd-id128/id128-util.c
Comment (daandemeyer) [inline src/libsystemd/sd-id128/id128-util.c]: This function should not do the symlink chasing. We should leave that to `id128_get_machine_at()`. This one should use `xopenat()` and then work on the file descriptor returned by that.

File: src/libsystemd/sd-id128/sd-id128.c
Comment (daandemeyer) [inline src/libsystemd/sd-id128/sd-id128.c]: This function is where we should use `chase()` (or a helper) to to symlink chasing. I would also rename `dir_fd` to `rfd` here to indicate that we see it as a root directory file descriptor.

File: src/libsystemd/sd-id128/sd-id128.c
Comment (daandemeyer) [inline src/libsystemd/sd-id128/sd-id128.c]: I would always open a file descriptor to the given root directory here to avoid accidental issues where we pass a relative path with `AT_FDCWD` and not get the result we expect.

File: src/libsystemd/sd-id128/id128-util.c
Comment (daandemeyer) [inline src/libsystemd/sd-id128/id128-util.c]: Similar here, this function should not do symlink chasing, we should instead add `id128_write_machine_at()` that does the symlink chasing and then calls this function to actually write the machine ID.

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
Some more comments.

File: src/basic/chase.c
Comment (yuwata) [reply src/basic/chase.c:576]: This is just for safety, and should not change any behavior. I extended the commit message.

If you do not like it, then I will drop the commit.

File: src/basic/chase.c
Comment (yuwata) [reply src/basic/chase.c:75]: - That's super inefficient first mixing the result from `chaseat()` with the (possibly adjusted) root path, and then splitting it again into the root part and sub part in `path_extract_filename()`.
- In `chase()`, the path to `root` directory may not be used as is. So, if we extend `path_extract_filename()`, then we need to sync the logic of root path handling in `chase()` and `path_extract_filename()`. Hence, the maintainability of the code becomes bad...

File: src/libsystemd/sd-id128/id128-util.c
Comment (yuwata) [reply src/libsystemd/sd-id128/id128-util.c]: OK. You mean you do not like 830e52caa2bf1a29f56cb93e7ed85acb1bda11c3. I will revert it and rework in that way.

File: src/shared/machine-id-setup.c
Comment (daandemeyer) [inline src/shared/machine-id-setup.c]: We should `chase_and_open()` "var/lib/dbus/machine-id" here and then just id128_read() the fd. WIth the current approach, if var/, lib/ or dbus/ are symlinks, we'll end up escaping the root directory since `ID128_NOFOLLOW` only applies to the final component of the path.

File: src/basic/chase.c
Comment (daandemeyer) [reply src/basic/chase.c:75]: - We have to iterate a few bytes and do one more allocation, that seems OK in terms of efficiency to me. Paths aren't long enough that iterating them is going to be noticeable in profiling
- I don't really follow the second point. If `CHASE_EXTRACT_FILENAME` is used, there's no extra logic needed by the caller to extract the filename. If it is not used, you get the full path and you can just use `path_extract_filename(root, p, &f)` to get the filename. That seems OK to me. If `chaseat()` is used, you can simply do `patch_extract_filename(NULL, p, &f)` to get the filename.

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
Almost OK, but I'm not still not entirely sure about adding chase_full().

**Comment by yuwata** [comment]
sd-id128 related commits are split out and already merged. Let's continue to discuss about `chase_full()` vs `CHASE_EXTRACT_FILENAME`.

**Comment by DaanDeMeyer** [comment]
So I think I still prefer adding an optional `root` argument to `path_extract_filename()` and not introducing all the full() variants to chase(). So let's wait for some more opinions from others before deciding which way to go.

**Comment by yuwata** [comment]
> So let's wait for some more opinions from others before deciding which way to go.

Yeah. I agree with that point.

**Comment by yuwata** [comment]
Superseded by #27918.

**Pull request closed without merging**
